sdk/core/azure-core/inc/azure/core/datetime.hpp

```diff
@@ -12,19 +12,161 @@
    */
   class DateTime {
   public:
-    /// A type that represents tick spans.
-    typedef uint64_t IntervalType;
-
-  private:
-    //  Number of seconds between 01-01-1970 and 01-01-1601.
-    static constexpr IntervalType WindowsToPosixOffsetSeconds = 11644473600LL;
+    /**
+     * @brief Represents time duration.
+     */
+    class Duration {
+      friend class DateTime;
+
+      int64_t m_100nsIntervals;
+      constexpr explicit Duration(int64_t intervalsOf100ns) : m_100nsIntervals(intervalsOf100ns) {}
+
+      static constexpr int64_t NanosecondResolution = 100;
+      static constexpr int64_t IntervalsOf100nsPerMicrosecond = 10;
+      static constexpr int64_t IntervalsOf100nsPerMillisecond
+          = 1000 * IntervalsOf100nsPerMicrosecond;
+      static constexpr int64_t IntervalsOf100nsPerSecond = 1000 * IntervalsOf100nsPerMillisecond;
+      static constexpr int64_t IntervalsOf100nsPerMinute = 60 * IntervalsOf100nsPerSecond;
+      static constexpr int64_t IntervalsOf100nsPerHour = 60 * IntervalsOf100nsPerMinute;
+
+    public:
+      constexpr Duration() : m_100nsIntervals(0) {}
+
+      static constexpr Duration FromHours(int32_t hours)
+      {
+        return Duration(hours * IntervalsOf100nsPerHour);
+      }
+
+      void constexpr AddHours(int32_t hours)
+      {
+        m_100nsIntervals += (hours * IntervalsOf100nsPerHour);
+      }
+
+      static constexpr Duration FromMinutes(int64_t minutes)
+      {
+        return Duration(minutes * IntervalsOf100nsPerMinute);
+      }
+
+      void constexpr AddMinutes(int64_t minutes)
+      {
+        m_100nsIntervals += (minutes * IntervalsOf100nsPerMinute);
+      }
+
+      static constexpr Duration FromSeconds(int64_t seconds)
+      {
+        return Duration(seconds * IntervalsOf100nsPerSecond);
+      }
+
+      void constexpr AddSeconds(int64_t seconds)
+      {
+        m_100nsIntervals += (seconds * IntervalsOf100nsPerSecond);
+      }
+
+      static constexpr Duration FromMilliseconds(int64_t milliseconds)
+      {
+        return Duration(milliseconds * IntervalsOf100nsPerMillisecond);
+      }
+
+      void constexpr AddMilliseconds(int64_t milliseconds)
+      {
+        m_100nsIntervals += (milliseconds * IntervalsOf100nsPerMillisecond);
+      }
+
+      static constexpr Duration FromMicroseconds(int64_t microseconds)
+      {
+        return Duration(microseconds * IntervalsOf100nsPerMicrosecond);
+      }
+
+      void constexpr AddMicroseconds(int64_t microseconds)
+      {
+        m_100nsIntervals += (microseconds * IntervalsOf100nsPerMillisecond);
+      }
+
+      static constexpr Duration FromNanoseconds(int64_t nanoseconds)
+      {
+        if (nanoseconds % NanosecondResolution != 0)
+        {
+          throw DateTimeException("Can't convert from nanoseconds");
+        }
+
+        return Duration(nanoseconds / NanosecondResolution);
+      }
+
+      void constexpr AddNanoseconds(int64_t nanoseconds)
```

Comment:
Add input validation for negative values?

Reply:
No, since we only have Add() method, the way you subtract is by providing negative value to add.

Reply:
I see. Gotcha. Do we have public API documentation explaining this?